### PR TITLE
DNS caching could be turned off, auto update of DNS caching

### PR DESCRIPTION
**Pull request by ludv1x**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

**Comment by jpiper** [comment]
Ah, didn't even know that ClickHouse had DNS caching. Would be useful to disable this in situations where the active endpoint for a service (e.g. an external dictionaries endpoint) is managed via DNS with a short TTL.

**Comment by ludv1x** [comment]
> Ah, didn't even know that ClickHouse had DNS caching. Would be useful to disable this in situations where the active endpoint for a service (e.g. an external dictionaries endpoint) is managed via DNS with a short TTL.

In the next release it will be possible to set `<disable_internal_dns_cache>1</disable_internal_dns_cache>` in the `config.xml` to turn off internal DNS cache.

**Comment by alexey-milovidov** [comment]
Pending:
- using hostnames in directory names of Distributed tables;
- using hostnames for names of nodes in ZK for distributed DDL.

File: dbms/src/Server/Server.cpp [hunk 1]
```diff
@@ -323,8 +324,17 @@ int Server::main(const std::vector<std::string> & /*args*/)
         global_context->setDDLWorker(std::make_shared<DDLWorker>(ddl_zookeeper_path, *global_context, &config(), "distributed_ddl"));
     }
 
-    /// Initialize a watcher updating DNS cache in case of network errors
-    DNSCacheUpdater dns_cache_updater(*global_context);
+    std::unique_ptr<DNSCacheUpdater> dns_cache_updater;
+    if (config().has("disable_internal_dns_cache") && config().getInt("disable_internal_dns_cache"))
```
Comment (alexey-milovidov) [inline dbms/src/Server/Server.cpp]: getBool?

File: dbms/src/Interpreters/Cluster.cpp [hunk 2]
```diff
@@ -94,7 +88,16 @@ String Cluster::Address::toString(const String & host_name, UInt16 port)
 
 String Cluster::Address::readableString() const
 {
-    return host_name + ':' + DB::toString(port);
+    String res;
+
+    /// If it looks like IPv6 address add braces to avoid ambiguity in ipv6_host:port notation
+    if (host_name.find_first_of(':') != std::string::npos && !host_name.empty() && host_name.back() != ']')
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/Cluster.cpp]: :+1:

File: dbms/src/Storages/MergeTree/BackgroundProcessingPool.cpp [hunk 3]
```diff
@@ -180,6 +181,7 @@ void BackgroundProcessingPool::threadFunction()
         catch (...)
         {
             tryLogCurrentException(__PRETTY_FUNCTION__);
+            DNSCacheUpdater::incrementNetworkErrors();
```
Comment (alexey-milovidov) [inline dbms/src/Storages/MergeTree/BackgroundProcessingPool.cpp]: Is it unfinished?
Why we increment network errors here?

File: dbms/src/Storages/MergeTree/BackgroundProcessingPool.cpp [hunk 4]
```diff
@@ -36,7 +37,7 @@ void BackgroundProcessingPool::TaskInfo::wake()
         std::unique_lock<std::mutex> lock(pool.tasks_mutex);
 
         auto next_time_to_execute = iterator->first;
-        TaskHandle this_task_handle = iterator->second;
+        auto this_task_handle = iterator->second;
```
Comment (alexey-milovidov) [inline dbms/src/Storages/MergeTree/BackgroundProcessingPool.cpp]: This change doesn't make anything better.

Comment (ludv1x) [reply dbms/src/Storages/MergeTree/BackgroundProcessingPool.cpp]: From appearance point of view it is better than `std::shared_ptr<BackgroundProcessingPoolTaskInfo> this_task_handle = iterator->second;`. (`TaskHandle` is unaccessable here after extracting `BackgroundProcessingPoolTaskInfo` from `BackgroundProcessingPool::Task`).

But it makes sense to use const references here.

File: dbms/src/Storages/MergeTree/BackgroundProcessingPool.cpp (hunk 3 above)
Comment (ludv1x) [reply dbms/src/Storages/MergeTree/BackgroundProcessingPool.cpp]: Background tasks [here](https://github.com/yandex/ClickHouse/pull/2141/files#diff-6d7b8a71e9cf645fa1e86f3e40f7ff25R178) could be affected by network errors also (for example fetching data from outdated IP address).
So, we need to count them also.

**Pull request merged**